File: LP-3/DAA/4 Nqueen/Nqueen.java

class Nqueen {
    public static boolean isSafe(int[][] board, int row, int col) {

        for (int i = 0; i < col; i++) { // row check kr
            if (board[row][i] == 1)
                return false;
        }

        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) { // upper-diagonal check kr
            if (board[i][j] == 1)
                return false;
        }

        for (int i = row, j = col; i < board.length && j >= 0; i++, j--) { // bottom-diagonal check kr
            if (board[i][j] == 1)
                return false;
        }

        return true;
    }

    public static boolean solveNQueenUtil(int[][] board, int col) {

        if (col >= board.length) {
            return true;
        }

        for (int i = 0; i < board.length; i++) {
            if (isSafe(board, i, col)) {
                board[i][col] = 1;
                printboard(board);
                System.out.println();
                if (solveNQueenUtil(board, col + 1)) {
                    return true;
                }
                board[i][col] = 0;

            }
        }
        return false;
    }

    public static void printboard(int board[][]) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board.length; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
    }

    public static void solveNQueen(int n) {
        int[][] board = new int[n][n];

        if (!solveNQueenUtil(board, 0)) {
            System.out.println("Solution does not exist");
        } else {
            printboard(board);
        }
    }

    public static void main(String args[]) {
        int n = 4;

        solveNQueen(n);
    }
}

File: nqueen.cpp
#include<iostream>
using namespace std;

bool isSafe(int board[8][8], int row, int col) {
    for (int i = 0; i < col; i++)
        if (board[row][i])
            return false;

    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
        if (board[i][j])
            return false;

    for (int i = row, j = col; i < 8 && j <8; i++, j++)
        if (board[i][j])
            return false;

    return true;
}
void solveNQueens(int board[8][8], int col,int N) {
 
    if (col == N) {
        // Print the solution
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++)
                cout << board[i][j] << " ";
            cout << endl;
        }
        cout << "-------------------\n";
        return;
    }


    for (int row = 0; row < N; row++) {

        if (isSafe(board, row, col)) {
            // Place the queen
            board[row][col] = 1;
            solveNQueens(board, col + 1,N);
            board[row][col] = 0;
        }
    }
}

int main(){

    int N = 8;
    int board[8][8] = {0};

    board[0][0] = 1;


    solveNQueens(board,1,N);

    return 0;
}